File: Commons/XML/src/main/java/com/greentree/commons/xml/parser/SAXXMLParser.java
package com.greentree.commons.xml.parser;

import com.greentree.commons.util.exception.WrappedException;
import com.greentree.commons.xml.XMLElement;
import org.w3c.dom.Document;
import org.w3c.dom.Element;
import org.w3c.dom.NodeList;
import org.w3c.dom.Text;
import org.xml.sax.InputSource;
import org.xml.sax.helpers.DefaultHandler;

import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.ParserConfigurationException;
import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.Map;

@Deprecated
public class SAXXMLParser {

    private static final DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();

    static {
    }

    private SAXXMLParser() {
    }

    public static XMLElement parse(final InputStream in) throws ParserConfigurationException, IOException {
        final DocumentBuilder builder = SAXXMLParser.factory.newDocumentBuilder();
        builder.setErrorHandler(new DefaultHandler());
        var is = new InputSource(in);
        final Document doc;
        try {
            doc = builder.parse(is);
        } catch (Throwable e) {
            throw new WrappedException(e);
        }
        return build(doc.getDocumentElement());
    }

    public static XMLElement build(final Element xmlElement) throws IOException, ParserConfigurationException {
        String name = xmlElement.getTagName();
        Collection<XMLElement> childrens = new ArrayList<>();
        final NodeList list = xmlElement.getChildNodes();
        for (int i = 0; i < list.getLength(); ++i) {
            var item = list.item(i);
            if (item instanceof Element) childrens.add(build((Element) item));
        }
        StringBuilder sbuilder = new StringBuilder();
        for (int i = 0; i < list.getLength(); ++i) {
            var item = list.item(i);
            if (item instanceof Text) sbuilder.append(item.getNodeValue());
        }
        String content = sbuilder.toString().replace("\n", "").replace("\t", "");
        Map<String, String> attributes = new HashMap<>();
        var map = xmlElement.getAttributes();
        for (int i = 0; i < map.getLength(); i++) {
            var a = map.item(i);
            var n = a.getNodeName();
            var v = a.getTextContent();
            attributes.put(n, v);
        }
        return new XMLElement(childrens, attributes, name, content);
    }

}

File: Commons/XML/src/main/java/com/greentree/commons/xml/parser/SAXXMLParser.kt
package com.greentree.commons.xml.parser

import com.greentree.commons.xml.XMLElement
import com.greentree.commons.xml.parser.MyXMLParser.replaceReference
import org.w3c.dom.Element
import org.w3c.dom.Text
import org.xml.sax.InputSource
import org.xml.sax.helpers.DefaultHandler
import java.io.IOException
import java.io.InputStream
import javax.xml.parsers.DocumentBuilder
import javax.xml.parsers.DocumentBuilderFactory
import javax.xml.parsers.ParserConfigurationException

@Deprecated("")
data object SAXXMLParser : XMLParser {

	override fun parse(inputStream: InputStream): XMLElement {
		val builder: DocumentBuilder = factory.newDocumentBuilder()
		builder.setErrorHandler(DefaultHandler())
		val `is` = InputSource(inputStream)
		val doc = builder.parse(`is`)
		return build(doc.getDocumentElement())
	}

	private val factory: DocumentBuilderFactory = DocumentBuilderFactory.newInstance()

	@Throws(IOException::class, ParserConfigurationException::class)
	private fun build(xmlElement: Element): XMLElement {
		val name = xmlElement.getTagName()
		val children = mutableListOf<XMLElement>()
		val list = xmlElement.getChildNodes()
		val builder = StringBuilder()
		for(i in 0 ..< list.getLength()) {
			val item = list.item(i)
			if(item is Element) children.add(build(item))
			if(item is Text) builder.append(item.getNodeValue().trim())
		}
		val attributes = HashMap<String, String>()
		val map = xmlElement.getAttributes()
		for(i in 0 ..< map.getLength()) {
			val a = map.item(i)
			val n = a.getNodeName()
			val v = a.getTextContent()
			attributes.put(n, v)
		}
		val content = replaceReference(builder.toString())
		return XMLElement(children, attributes, name, content)
	}
}
